keras/layers/convolutional_recurrent.py

```diff
@@ -10,167 +10,455 @@
-                  'padding': self.padding,
-                  'data_format': self.data_format,
-                  'dilation_rate': self.dilation_rate,
-                  'return_sequences': self.return_sequences,
-                  'go_backwards': self.go_backwards,
-                  'stateful': self.stateful}
-        base_config = super(ConvRecurrent2D, self).get_config()
-        return dict(list(base_config.items()) + list(config.items()))
+        # helper function
+        def get_tuple_shape(nb_channels):
+            result = list(state_shape)
+            if self.cell.data_format == 'channels_first':
+                result[1] = nb_channels
+            elif self.cell.data_format == 'channels_last':
+                result[3] = nb_channels
+            else:
+                raise KeyError
+            return tuple(result)
 
+        # initialize state if None
+        if self.states[0] is None:
+            if hasattr(self.cell.state_size, '__len__'):
+                self.states = [K.zeros(get_tuple_shape(dim))
+                               for dim in self.cell.state_size]
+            else:
+                self.states = [K.zeros(get_tuple_shape(self.cell.state_size))]
+        elif states is None:
+            if hasattr(self.cell.state_size, '__len__'):
+                for state, dim in zip(self.states, self.cell.state_size):
+                    K.set_value(state, np.zeros(get_tuple_shape(dim)))
+            else:
+                K.set_value(self.states[0],
+                            np.zeros(get_tuple_shape(self.cell.state_size)))
+        else:
+            if not isinstance(states, (list, tuple)):
+                states = [states]
+            if len(states) != len(self.states):
+                raise ValueError('Layer ' + self.name + ' expects ' +
+                                 str(len(self.states)) + ' states, '
+                                                         'but it received ' + str(len(states)) +
+                                 ' state values. Input received: ' +
+                                 str(states))
+            for index, (value, state) in enumerate(zip(states, self.states)):
+                if hasattr(self.cell.state_size, '__len__'):
+                    dim = self.cell.state_size[index]
+                else:
+                    dim = self.cell.state_size
+                if value.shape != get_tuple_shape(dim):
+                    raise ValueError('State ' + str(index) +
+                                     ' is incompatible with layer ' +
+                                     self.name + ': expected shape=' +
+                                     str(get_tuple_shape(dim)) +
+                                     ', found shape=' + str(value.shape))
+                # TODO: consider batch calls to `set_value`.
+                K.set_value(state, value)
 
-class ConvLSTM2D(ConvRecurrent2D):
-    """Convolutional LSTM.
+    @property
+    def losses(self):
```

Comment:
I believe you don't need this method, the inherited one should work fine.

Reply:
The issue is that removing this function causes the test to fail:
```
layer = convolutional_recurrent.ConvLSTM2D(**kwargs)
                    layer.build(inputs.shape)
                    assert len(layer.losses) == 3
                    assert layer.activity_regularizer
                    output = layer(K.variable(np.ones(inputs.shape)))
>                   assert len(layer.losses) == 4
E                   AssertionError: assert 3 == 4
E                    +  where 3 = len([<tf.Tensor 'add:0' shape=() dtype=float32>, <tf.Tensor 'add_1:0' shape=() dtype=float32>, <tf.Tensor 'add_2:0' shape=() dtype=float32>])
E                    +    where [<tf.Tensor 'add:0' shape=() dtype=float32>, <tf.Tensor 'add_1:0' shape=() dtype=float32>, <tf.Tensor 'add_2:0' shape=() dtype=float32>] = <keras.layers.convolutional_recurrent.ConvLSTM2D object at 0x00000262BA0BFBA8>.losses
```

I'm going to investigate and see what can be done.

Reply:
So, after some research, I believe that we need either to keep it or change the `losses` property in the `RNN` class. Let me explain:

This test verifies that the activity regularizer loss can be obtained from the layer. If we fall back to the `RNN` loss, then this code get executed:

```python
@property
def losses(self):
    if isinstance(self.cell, Layer):
        return self.cell.losses
    return []
```
which implies that the cell contains all the losses we are interested in. But this is a contradiction with the tests that I saw in `recurrent_test.py` notably, those lines:

```python
layer = layer_class(units, return_sequences=False, weights=None,
                        input_shape=(timesteps, embedding_dim),
                        activity_regularizer='l2')
assert layer.activity_regularizer
x = K.variable(np.ones((num_samples, timesteps, embedding_dim)))
layer(x)
assert len(layer.cell.get_losses_for(x)) == 0
assert len(layer.get_losses_for(x)) == 1
```

In the end, I see no mention anywhere in `LSTMCell` of the property `activity_regularizer` leading me to believe that the cell does usually not hold this loss. 
So this is strange. I made a last commit modifying the `losses` attribute of `RNN` because I think this is an expected behavior to get the loss relative to the activity regularizer when calling the attribute. Please tell me if this is what you think is right.